app/services/post_alerter.rb

```diff
@@ -140,7 +140,8 @@ def notify_first_post_watchers(post, user_ids)
     users = User.where(id: user_ids)
 
     DiscourseEvent.trigger(:before_create_notifications_for_users, users, post)
-    users.each do |user|
+    users.pluck(:id).each do |user_id|
```

Comment:
Maybe `find_in_batches` would be better here?

Reply:
It would probably be more efficient, but find_in_batches can be [affected by race conditions](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_in_batches) (e.g. if a user at the beginning of the list is deleted while the notifications are being sent, it could cause a user near the end of the list to be double-notified)

Loading one-by-one matches the existing behaviour of [`PostAlerter#notify_post_users`](https://github.com/discourse/discourse/blob/master/app/services/post_alerter.rb#L614-L618)

Reply:
That's a nice solution. I think we should do it here and elsewhere.

Reply:
:+1: updated `notify_first_post_watchers` to use batches, and also `#notify_post_users`